tests/pthread/test_pthread_cancel_cond_wait.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <pthread.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <unistd.h>
+#include <errno.h>
+#include <emscripten.h>
+#include <emscripten/threading.h>
+
+pthread_barrier_t barrier;
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t condvar = PTHREAD_COND_INITIALIZER;
+
+int th_cancelled = 0;
+
+volatile int res = 43;
+static void cleanup_handler(void *arg)
+{
+  EM_ASM(out('Called clean-up handler with arg ' + $0), arg);
+  int a = (int)arg;
+  res -= a;
+
+  pthread_mutex_unlock(&mutex);
+  pthread_barrier_wait(&barrier);
+}
+
+static void *thread_start(void *arg)
+{
+  pthread_cleanup_push(cleanup_handler, (void*)42);
+  EM_ASM(out('Thread started!'));
+  pthread_mutex_lock(&mutex);
+  pthread_barrier_wait(&barrier);
+
+  int ret = 0;
+  do {
+    EM_ASM(out('Waiting on conditional variable'););
+    ret = pthread_cond_wait(&condvar, &mutex);
+  } while (ret == 0 && th_cancelled == 0);
+
+  if (ret != 0) {
+    EM_ASM(out('Cond wait failed ret: ' +  $0);, ret);
+  }
+
+  res = 1000; // Shouldn't ever reach here.
+  pthread_cleanup_pop(0);
+
+  pthread_mutex_unlock(&mutex);
+  pthread_barrier_wait(&barrier);
+  return NULL;
+}
+
+pthread_t thr;
+
+int main()
+{
+  if (!emscripten_has_threading_support())
+  {
+#ifdef REPORT_RESULT
+    REPORT_RESULT(1);
+#endif
+    printf("Skipped: Threading is not supported.\n");
+    return 0;
+  }
+
+  pthread_barrier_init(&barrier, nullptr, 2);
+
+  int s = pthread_create(&thr, NULL, thread_start, (void*)0);
+  assert(s == 0);
+  EM_ASM(out('Thread created'););
+
+  pthread_barrier_wait(&barrier);
+
+  // Lock mutex to ensure that thread is waiting
+  pthread_mutex_lock(&mutex);
+
+  EM_ASM(out('Canceling thread..'););
+  s = pthread_cancel(thr);
+  assert(s == 0);
+  th_cancelled = 1;
+  pthread_mutex_unlock(&mutex);
+
+  EM_ASM(out('Main thread waitnig for side-thread'););
+  pthread_barrier_wait(&barrier);
+  pthread_barrier_destroy(&barrier);
+
+  EM_ASM(out('Test finished result: ' + $0);, res);
+#ifdef REPORT_RESULT
+  REPORT_RESULT(res);
+#endif
+  return res != 1;
+}
```

Comment:
I'm curious how if this new test is testing similar things to the newly enabled posixtestsuite tests?

Reply:
Actually yes, this test is supposed to check similar scenario to one tested in pthread_cond_wait_2_3 test. Although in this test scenario itself is simplified compared to test from posixtestsuite. Main purpose of this test was to check cancelling thread waiting on cv scenario without adding posixtestsuite.

I don't know if posixtestsuite is run by CI but if so then this test can be removed. Do you think it is better to keep this tests or remove it as redundant?